src/jaegertracing/net/IPAddress.cpp

```diff
@@ -15,38 +15,92 @@
  */
 
 #include "jaegertracing/net/IPAddress.h"
+#include "jaegertracing/net/Socket.h"
 
-#include <ifaddrs.h>
 #include <sys/types.h>
 
+#ifndef WIN32
+#include <ifaddrs.h>
+#endif
+
 namespace jaegertracing {
 namespace net {
+
 namespace {
 
-struct IfAddrDeleter : public std::function<void(ifaddrs*)> {
-    void operator()(ifaddrs* ifAddr) const
+template <typename T>
+struct CDeleter : public std::function<void(T*)> {
+    void operator()(T* ifAddr) const
     {
         if (ifAddr) {
+#ifdef WIN32
+            free(ifAddr);
+#else
             ::freeifaddrs(ifAddr);
+#endif
         }
     }
 };
 
 }  // anonymous namespace
 
+#ifdef WIN32
+
 IPAddress IPAddress::localIP(int family)
 {
-    return localIP([family](const ifaddrs* ifAddr) {
-        return ifAddr->ifa_addr != nullptr &&
-               ifAddr->ifa_addr->sa_family == family;
-    });
+    DWORD size = 15032;
+    DWORD rv = GetAdaptersAddresses(
+        AF_UNSPEC, GAA_FLAG_INCLUDE_PREFIX, NULL, NULL, &size);
+    if (rv != ERROR_BUFFER_OVERFLOW) {
+        // GetAdaptersAddresses() failed...
+        return IPAddress();
+    }
+    std::unique_ptr<IP_ADAPTER_ADDRESSES, CDeleter<IP_ADAPTER_ADDRESSES>>
+        adapter_addresses((IP_ADAPTER_ADDRESSES*)malloc(size));
+
+    rv = GetAdaptersAddresses(
+        AF_UNSPEC, GAA_FLAG_INCLUDE_PREFIX, NULL, &*adapter_addresses, &size);
+    if (rv != ERROR_SUCCESS) {
+        // GetAdaptersAddresses() failed...
+        return IPAddress();
+    }
+
+    for (PIP_ADAPTER_ADDRESSES aa = &*adapter_addresses; aa != NULL;
+         aa = aa->Next) {
+        for (PIP_ADAPTER_UNICAST_ADDRESS ua = aa->FirstUnicastAddress;
+             ua != NULL;
+             ua = ua->Next) {
+            if (family == ua->Address.lpSockaddr->sa_family) {
+                char buf[BUFSIZ] = { 0 };
+                getnameinfo(ua->Address.lpSockaddr,
+                            ua->Address.iSockaddrLength,
+                            buf,
+                            sizeof(buf),
+                            NULL,
+                            0,
+                            NI_NUMERICHOST);
+
+                return IPAddress(*(ua->Address.lpSockaddr),
+                                 (family == AF_INET) ? sizeof(::sockaddr_in)
+                                                     : sizeof(::sockaddr_in6));
+            }
+        }
+    }
+
+    return IPAddress();
 }
 
-IPAddress IPAddress::localIP(std::function<bool(const ifaddrs*)> filter)
+#else
+
+#include <ifaddrs.h>
```

Comment:
I'd rather not include this mid-file. Can you please use `#ifndef WIN32` to wrap this include after the system includes at the top?